Add lecture participation tracker using GitHub Actions (#2459)

bjornthiberg opened the pull request:
# Description

This PR introduces a script and a GitHub Actions workflow to track lecture participation.

The structure of the script and workflow YAML file is based on the existing workflow for updating the Statistics of Student Registrations. It has a similar structure, can be run locally in the same way, and uses the same dependencies.

How it's supposed to work:

GitHub actions workflow
- The workflow is triggered on new issue comments, and proceeds if the issue title is “Lecture participation tracker 2024”.
- track_participation.py is run

Python script (`track_participation.py`):
- Fetches all comments to the issue from the GitHub API
- If the author is a collaborator, or the comment is outside the commenting window (5 hours from the start of each lecture), the comment is ignored.
- Valid comments are stored in a dictionary, with keys for each student who has commented.
- The script generates a markdown table, similar to the one for tracking tasks, which displays the lectures each student has attended.
- The issue description is updated with this markdown table and a timestamp.

I have tested that (in a copy of the repo):
- the workflow triggers correctly when a new comment is added to the tracker.
- comments made by collaborators are ignored as expected.
- comments made outside of the allowed window are not counted.
- valid comments update the participation tracker issue description.
- timezones work as expected (don't quote me on this)

I have not tested for more than two unique students/commenters, as would be the case in "production".

Updating for newer years should only require changing the lecture dates and start hours in the JSON config and the issue number in the YAML file. 

Closes #2459

bjornthiberg committed:
feat: add lecture participation tracker using GitHub Actions

algomaster99 commented on .github/workflows/lecture_participation.yml:
```diff
@@ -0,0 +1,36 @@
+
+jobs:
+  track-participation:
+    if: github.event.issue.title == 'Lecture participation tracker 2024'
```
Since the issue title is mutable, this could break. I suggest comparing equality with the issue number instead.

bjornthiberg replied:
I thought about how to implement this. One could do it as in [assignment-statistics.yml](https://github.com/KTH/devops-course/blob/87ecf8925262aab78ed8098b1438abadd0a2ee1e/.github/workflows/assignment-statistics.yml), where it's set as an environmental variable and then checked in the python script. The problem with that is that it will unnecessarily continue the CI runner, set up Python, install dependencies etc., even if it's not the correct issue. And I suppose changing a number in the workflow is not more work than changing a repo secret.

So I'll probably just do

`if: github.event.issue.number == 2370`

bjornthiberg replied:
Addressed in commit a0a9a86

algomaster99 replied:
I am okay with this approach.

algomaster99 commented on .github/workflows/lecture_participation.yml:
```diff
@@ -0,0 +1,36 @@
+    if: github.event.issue.title == 'Lecture participation tracker 2024'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

We use [latest versions](https://github.com/marketplace/actions/checkout) which reminds me that we need another issue for integrating [renovate](https://github.com/renovatebot/renovate). :)

bjornthiberg replied:
Woops, will fix.

needs to be updated in [assignment-statistics.yml](https://github.com/KTH/devops-course/blob/87ecf8925262aab78ed8098b1438abadd0a2ee1e/.github/workflows/assignment-statistics.yml), along with setup python

bjornthiberg replied:
Addressed in commit ab6b4e5

algomaster99 replied:
Thanks @bjornthiberg ! You can always use "commit suggestion" in order to approve these changes instead of committing separately.
![image](https://github.com/user-attachments/assets/590a5ff6-c4d7-461e-a7d8-4e0790ae482a)

algomaster99 commented on .github/workflows/lecture_participation.yml:
```diff
@@ -0,0 +1,36 @@
+      - uses: actions/checkout@v2
+      
+      - name: Setup Python
+        uses: actions/setup-python@v2.2.1
```
```suggestion
        uses: actions/setup-python@v5.2.0
```

bjornthiberg replied:
Addressed in commit ab6b4e5

algomaster99 commented on .github/workflows/lecture_participation.yml:
```diff
@@ -0,0 +1,36 @@
+      - name: Setup Python
+        uses: actions/setup-python@v2.2.1
+        with:
+          python-version: '3.x'
```
Pin a specific python version. See the importance of [pinning](https://docs.renovatebot.com/dependency-pinning/).

bjornthiberg replied:
Addressed in commit ab6b4e5

algomaster99 commented on .github/workflows/lecture_participation.yml:
```diff
@@ -0,0 +1,36 @@
+          if [ -f ./tools/requirements.txt ]; then pip install -r ./tools/requirements.txt; fi
+
+      - name: Set timezone to Europe/Stockholm
+        run: sudo timedatectl set-timezone Europe/Stockholm
```
Why do you need this here? Your python script fetches the comments which have the time in UTC so python will have libraries to convert times, right?

bjornthiberg replied:
Could definitely be done in the Python script instead of using system time, in turn set in the workflow. It's another remnant from assignment-statistics.yml, but probably cleaner to just do it in the script.

bjornthiberg replied:
Addressed in e8d3079

algomaster99 commented on tools/track_participation.py:
```diff
@@ -0,0 +1,224 @@
+
+    for comment in tracking_issue.get_comments():
+        author = comment.user.login
+        comment_time = comment.created_at.astimezone()
```
Here you could convert it to `Europe/Stockholm` timezone instead in the workflow file, right?

bjornthiberg replied:
Addressed in e8d3079

algomaster99 commented on tools/track_participation.py:
```diff
@@ -0,0 +1,224 @@
+LECTURE_DATES_TO_START_TIME = {
+    "2024-09-04": 13, "2024-09-11": 13, "2024-09-18": 13, "2024-09-25": 13,
+    "2024-10-02": 13, "2024-10-09": 13
+}
```
This is configuration information. Seggrate this from the main script.

bjornthiberg replied:
One solution in 46064c3. 

LMK if having a separate JSON file is not what you wanted, could be in another config.py file as well.

bjornthiberg replied:
(small fix in 79ef6a4)

algomaster99 replied:
I am fine with the `JSON` file.

algomaster99 reviewed (changes requested):
Hi! Thank you so much for the wonderful PR :heart: Please see the changes that I have suggested.

1. What is the default option in your script? `--printPlain`, `--printMarkdown`, `--publish`? Running it without output any options simply exits the script after 3 seconds with code `0`.
3. Do you agree to support maintenance of this script until the end of the course? :) I ask this because there could be a high chance that I merge this PR and then action breaks so I would need your help to fix it :D 


@monperrus @Deee92 @javierron @sofiabobadilla example:
```
+-------+---------------------+-----------------------------+---------------------+
| Index |     Student Name    | Number of Lectures Attended | Lecture(s) attended |
+-------+---------------------+-----------------------------+---------------------+
|   1   |       RafDevX       |              2              |        L2 L3        |
|   2   |       Glassar       |              1              |          L2         |
|   3   |    diogotcorreia    |              2              |        L2 L3        |
|   4   |   MateusMarinheiro  |              2              |        L2 L3        |
|   5   |     bjornthiberg    |              1              |          L2         |
|   6   |       t-sorger      |              2              |        L2 L3        |
|   7   |      jackma-00      |              2              |        L2 L3        |
|   8   |    randomicecube    |              2              |        L2 L3        |
|   9   |        dvavd        |              2              |        L2 L3        |
|   10  |       Tian0602      |              2              |        L2 L3        |
|   11  |      Atheer2104     |              2              |        L2 L3        |
|   12  |      arejula27      |              2              |        L2 L3        |
|   13  |       d-melita      |              2              |        L2 L3        |
|   14  |        M1l0d        |              2              |        L2 L3        |
|   15  | CoordinatesNotFound |              2              |        L2 L3        |
|   16  |    marcocampione    |              2              |        L2 L3        |
|   17  |     Stellariser     |              2              |        L2 L3        |
|   18  |        NoelMT       |              2              |        L2 L3        |
|   19  |     pablo-freire    |              1              |          L2         |
|   20  |  GustavHenningsson  |              1              |          L2         |
|   21  |     Pesteves2002    |              2              |        L2 L3        |
|   22  |     muhammad1928    |              2              |        L2 L3        |
|   23  |       Uqqasha       |              2              |        L2 L3        |
|   24  |       Flopalot      |              1              |          L2         |
|   25  |      bepp-boop      |              1              |          L3         |
|   26  |   Alexanderliu2002  |              1              |          L3         |
|   27  |     laullaurado     |              1              |          L3         |
|   28  |       Jakebobs      |              1              |          L3         |
|   29  |     einbergisak     |              1              |          L3         |
|   30  |      Samkth123      |              1              |          L3         |
|   31  |        kthfre       |              1              |          L3         |
|   32  |      peremateu      |              1              |          L3         |
|   33  |        HexuL        |              1              |          L3         |
+-------+---------------------+-----------------------------+---------------------+
```

bjornthiberg commented:
Hi @algomaster99, thanks for the feedback!

I just did a force push to get the PR in line with the task structure, by changing the description and adding a task readme file (did not change the python script or YAML file). 

I also added a short documentation to the /tools/README.md file, didn't see that earlier.

1. No default option, it just does the fetching but prints an empty string. Doesn't make a lot of sense, I'll make it print plain text as default and remove it as an option instead.
2. Of course!

I'll get on your other comments asap.

bjornthiberg committed:
use issue number instead of title

bjornthiberg committed:
use newest versions of checkout and setup-python, pin python 3.11.8

bjornthiberg committed:
set timezone in script instead of workflow file

bjornthiberg committed:
make default behavior be print in plaintext

bjornthiberg committed:
Move configuration to external JSON file

bjornthiberg committed:
fix: use correct file path for config

bjornthiberg commented:
@algomaster99

I have addressed all your comments now I think! Let me know if I missed something or anything else needs to be adjusted :)

I have tested the code again, and the CI job [on my test repo](https://github.com/bjornthiberg/devops_course_test/actions/runs/10833973286/job/30062022237).

algomaster99 commented on contributions/open-source/bthiberg/README.md:
```diff
@@ -0,0 +1,53 @@
+# Assignment Proposal
```
Thanks for the proposal, but could you please make a separate PR about this? We take the task and the actual contribution separately.

bjornthiberg replied:
That makes sense, I have removed the proposal from this PR in dcdb3fd and submitted a separate one (#2480)

algomaster99 commented on tools/README.md:
```diff
@@ -58,3 +58,32 @@ This script is used to grade each student according to the number of task comple
 ...
+
+
+## track_participation.py
```
I almost forgot about suggesting this change, but I am happy you noticed! Could you please also add a heading about the prerequisites that should be updated every year?

bjornthiberg replied:
addressed in 71c9c9f

algomaster99 commented on tools/track_participation.py:
```diff
@@ -0,0 +1,215 @@
+REPO_FULLNAME = os.getenv("REPO_FULLNAME")
+ISSUE_NUMBER = os.getenv("TRACKER_ISSUE_NUMBER")
+
+config = json.load(open("tools/track_participation_config.json"))
```
```suggestion
config = json.load(open("track_participation_config.json"))
```

I think this should be sufficient?

bjornthiberg replied:
It will work when running locally from the tools folder, but the workflow will not find the file as it runs Python in the repo's root dir. 

[See this failed run in the test repo.](https://github.com/bjornthiberg/devops_course_test/actions/runs/10833877493/job/30061700682)

Could change that in the workflow file, but I thought this was cleaner.

algomaster99 replied:
> Could change that in the workflow file, but I thought this was cleaner.

I think otherwise because the this path is only true with respect to the CI job so it should be in workflow file only.

algomaster99 commented on tools/track_participation.py:
```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Filename: track_participation.py
```
May I ask what these headers are for? Is it required for some plugin that you use? I am curious.

bjornthiberg replied:
Just something I saw in the other python scripts in the repo. I believe the first line is for encoding, but is generally a remnant from python 2 as 3+ uses utf-8 by default. Second line is just for readability I suppose.

algomaster99 commented on tools/track_participation.py:
```diff
@@ -0,0 +1,215 @@
+    lecture_start_hour = LECTURE_DATES_TO_START_TIME[lecture_date_str]
+
+    allowed_period_start = comment_time.replace(hour=lecture_start_hour, minute=0, second=0, microsecond=0)
+    allowed_period_end = allowed_period_start + timedelta(hours=COMMENTING_DURATION_HOURS)
```
Sorry if I being too nitpicky here, but I would request you to have two variables here instead of one `COMMENT_DURATION_HOURS`. One which is the actual duration and the other which is the leeway. This would be my personal opinion in terms of code readability/understandability.

bjornthiberg replied:
Agreed! addressed in ef44d53

algomaster99 reviewed (changes requested):
> did a force push to get the PR in line with the task structure, by changing the description and adding a task readme file

Always create a PR from a different branch. You can avoid rebasing in this way.

> Of course!

Thank you!

I think we are almost there! Just a few more changes.

algomaster99 commented on contributions/open-source/bthiberg/README.md:
```diff
@@ -0,0 +1,53 @@
+
+I have not tested for more than two unique students/commenters, as would be the case in "production".
+
+Updating for newer years should only require changing the lecture dates constant in the Python code and the issue title in the YAML file. 
```
```suggestion
Updating for newer years should only require changing the lecture dates constant in the Python code and the issue number in the YAML file. 
```

bjornthiberg commented:
> Always create a PR from a different branch. You can avoid rebasing in this way.
 
Damnit, didn't think about this since it's a fork. Could change now but probably more trouble than it is worth since rebasing will not be very complicated.

> I think we are almost there! Just a few more changes.

On it!

bjornthiberg committed:
remove task proposal

bjornthiberg committed:
add required yearly updates to documentation

bjornthiberg committed:
separate commenting duration into lecture and  leeway durations

bjornthiberg commented:
@algomaster99 I've addressed the requested changes. Let me know if there’s anything else needed or if you spot any issues! :)

I suppose my commits should be squashed and rebased onto the current 2024 branch. Lmk if I should do that and force push to this PR, to make merging easier.

algomaster99 reviewed (changes requested):
Final change and I will merge it once you resolve it :)

bjornthiberg committed:
update workflow to run in tools dir, get config from current dir

algomaster99 commented:
Thanks for the quick fix! Let's give it a run, shall we? I am looking forward to it.

bjornthiberg committed:
fix: use timezone in "last changed" timestamps

bjornthiberg commented:
I squeezed in another quick fix! The timestamp ended up being incorrect with the timezone change, should be fixed now in 5c90115

algomaster99 commented:
> The timestamp ended up being incorrect with the timezone change, should be fixed now in https://github.com/KTH/devops-course/commit/5c90115b36fe587ebc92017071e036e1fc114884

Awesome! I did not notice it. Thanks!

algomaster99 commented:
Are we ready to merge? :smile:

algomaster99 commented:
Merging! Let's hope nothing breaks :D

algomaster99 commented:
Great work! One cool thing I immediately realised was that the table could be sorted by the number of lectures attended. Could you also propose a PR for that? I hope it is not too much effort.

bjornthiberg commented:
Should be easy! Will do next week

algomaster99 commented:
Thanks!
